url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/JGraphExampleLayoutCache.java

```diff
@@ -0,0 +1,393 @@
+						if (evt.getPropertyName().equals(
+								JGraphLayoutProgress.PROGRESS_PROPERTY)) {
+							int newValue = Integer.parseInt(String.valueOf(evt
+									.getNewValue()));
+							progressMonitor.setProgress(newValue);
+						} else if (evt.getPropertyName().equals(
+								JGraphLayoutProgress.MAXIMUM_PROPERTY)) {
+							int newValue = Integer.parseInt(String.valueOf(evt
+									.getNewValue()));
+							progressMonitor.setMaximum(newValue);
+						}
+
+						// Checks isCancelled as the progress
+						// monitor has no listener
+						if (progressMonitor.isCanceled())
+							progress.setStopped(true);
+					}
+				};
+
+				Thread layoutThread = new Thread() {
+					public void run() {
+						progress
+								.addPropertyChangeListener(propertyChangeListener);
+						// long t1 = System.currentTimeMillis();
+						try {
+							facade.run(layout, false);
+						} catch(Exception e) {
+						    e.printStackTrace();
+						}
+						SwingUtilities.invokeLater(new Runnable() {
+							public void run() {
+								progress
+										.removePropertyChangeListener(propertyChangeListener);
+
+								// After the layout terminates...
+								boolean ignoreResult = progressMonitor
+										.isCanceled();
+								progressMonitor.close();
+
+								// If not canceled process result
+								if (!ignoreResult)
+									processResult(facade, nestedMap);
+
+								// Else just repeat the original call
+								// which will also release the lock
+								else {
+									edit(nestedMap);
+								}
+							}
+						});
+					}
+				};
+				// While the layout runs...
+				layoutThread.start();
+			} else {
+
+				// Immediately runs and processes the result
+				facade.run(layout, false);
+				processResult(facade, nestedMap);
+			}
+		}
+	}
+
+	/**
+	 * Processes the result of the layout.
+	 */
+	protected void processResult(JGraphFacade facade, Map nestedMap) {
+		boolean flushToOrigin = false;
+		if (layoutFlushOrigin) {
+			flushToOrigin = true;
+		}
+		Map layoutResult = facade.createNestedMap(layoutIgnoreGrid, flushToOrigin);
+
+		// Remove changed bounds and points from the attributes
+		if ((layoutFlushOrigin || layoutMoveSelection) && nestedMap != null) {
+			layoutResult = GraphConstants.merge(layoutResult, new Hashtable(
+					nestedMap));
+		} else {
+			layoutResult = GraphConstants.merge(nestedMap, layoutResult);
+		}
+
+		// Morph into new positions only if additional cells found in layout
+		int manualCount = 0;
+		if (nestedMap != null)
+			manualCount = facade.getVertices(nestedMap.keySet(), false).size();
+
+		if (morphing && layoutResult != null
+				&& (nestedMap == null || manualCount <= layoutResult.size())) {
+			graph.morph(layoutResult, (nestedMap != null) ? nestedMap.keySet()
+					: null);
+		} else {
+			edit(layoutResult, null, null, null);
+		}
+	}
+
+	public Rectangle2D getBounds(Object cell, Map nested) {
+		Map attrs = (nested != null) ? (Map) nested.get(cell) : null;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;cell&#39; (index: 0 | type: Object). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;nested&#39;. Other methods validating the parameter: &#39;JGraphExampleLayoutCache.isMoveable&#39;. The same parameter type is validated here for example: url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/components/labels/MultiLineVertexView.java:54. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).